### DISCUSSION TITLE
Copy explicit `\t` (tab) with tab, not spaces

### DISCUSSION THREAD

**Discussion by daniilr** [ghostty-org/ghostty#3486] [Feature Requests, Ideas] [answered]
I am running Ghostty 1.0 on a mac. My shell is zsh. I was able to reproduce the issue in bash as well. The default Terminal.app works good. 

https://github.com/user-attachments/assets/cc5deea1-fce1-4bad-a800-e4c92f339d81
`➜  ~ echo -e 'test\u9test'
test    test
`

Is this behaivour expected?

**Comment by qwerasd205** [comment]
Yes this is expected behavior. It's not actually 4 spaces, it's simply aligning to the next tabstop. Perhaps it would be a good quality of life feature to encode the tabs as state in a way that lets the user copy them as tabs instead of spaces though.

**Comment by mitchellh** [comment]
> Perhaps it would be a good quality of life feature to encode the tabs as state in a way that lets the user copy them as tabs instead of spaces though.

Interesting.

Converting to a feature request discussion and renaming. I'd be curious next what other terminals do here. I'm willing to do something different, just curious!

**Reply by daiyam** [reply]
TE which allow to get back the tab character with copy:
- Terminal.app
- iterm2
- Kitty
- Alacritty
- TE based on VTE (most linux TE)
- Warp

TE which don't:
- xterm.js
- Git BASH (from Git for Windows)

About xterm.js, I did make a PR but was rejected and a discussion about that feature turned out to be a philosophical debate about the tab character.

Edit: add Warp

**Reply by 00-kat** [reply]
N.B.: Git Bash is actually MinTTY, which is what ships with MSYS2.

**Comment by indahazzard** [comment]
I've updated .editorconfig file to set indent_style to tab like this:
```.editorconfig
indent_style = tab
```
and this helped, not sure if it's your case.

**Reply by daniilr** [reply]
This will probably help for nvim when it saves a file, what I am trying to do is to copy and paste right from terminal

**Comment by mitchellh** [comment]
Okay, further question since other TEs do this. If I do the following sequence, what happens:

1. Cursor at x=0,y=0
2. Print "\tHello"
3. Cursor at x=1
4. Print " " (space)
5. Copy line

Does writing the one "space" turn the entire tab into spaces? Since tabs are translated into spaces for a terminal, I'm just curious how these sorts of edge cases are handled.

I'll write a shell script at some point to research this further.

**Reply by daiyam** [reply]
> If I do the following sequence, what happens:

Short answer: I don't know. 
I see the issue and I will try to look into it. 

Here my goto script to test the tab support: `printf "\033[48;5;1mred\t - \tred\033[0mnormal\n"`
(it's include some color tests but that's very low priority).

**Reply by daniilr** [reply]
It seems that mac Terminal.app keeps both tab and space character

**Reply by Flimm** [reply]
I'm pretty sure that the correct behaviour is to keep space characters as is, and to keep tab characters as is. I don't expect my terminal to run `sed` on the output of commands (except for things like ANSI colour escape sequences).

I'm not sure that the command posted by @daiyam does test support for tab characters properly. We need to check that characters are copied correctly into the clipboard, regardless of which colour they are dispayed in.

These terminal emulators copy tab characters properly in my testing on Linux:

- Kitty
- GNOME Terminal
- GNOME Console
- Terminator
- Tilix
- XFCE Terminal

These don't:

- Ghostty
- Konsole
- xterm
- alacritty
- From my recollection, Windows Terminal also mangles tab characters and replaces them with spaces on Windows

**Reply by daiyam** [reply]
> I'm not sure that the command posted by @daiyam does test support for tab characters properly.

I agree

> I'm pretty sure that the correct behaviour is to keep space characters as is, and to keep tab characters as is. I don't expect my terminal to run `sed` on the output of commands (except for things like ANSI colour escape sequences).
> We need to check that characters are copied correctly into the clipboard, regardless of which colour they are dispayed in.

I also agree. We only need the tab characters in the copy value.

> alacritty

I think that's depend of the platform.

**Reply by mitchellh** [reply]
> I'm pretty sure that the correct behaviour is to keep space characters as is, and to keep tab characters as is. I don't expect my terminal to run sed on the output of commands (except for things like ANSI colour escape sequences).

The problem is the terminal doesn't see a tab character. Tab is a control character that aligns whitespace to the nearest tabstop. You can see this in the most canonical terminal: xterm. xterm itself copies spaces. It requires special handling to convert the spaces BACK to a tab control character. 

(I agree with this behavior, just noting its not obvious)

**Comment by Flimm** [comment]
I regularly copy [TSV](https://en.wikipedia.org/wiki/Tab-separated_values) file output from the terminal and I paste it into Google Sheets or LibreOffice Calc. In Kitty, GNOME Terminal, and others, the tab characters are preserved, and the cells get slotted into the spreadsheet correctly. It's a nice feature in the spirit of Ghostty's goals of integrating well with the rest of the desktop environment. Sadly, I can't get this nice intergation with Ghostty, since Ghostty replaces tab characters with space characters. Google Sheets and LibreOffice Calc (rightly) don't try to handle spaces as cell separators, but they do know how to interpret tab characters.

Expected output (incudes a tab character):

```
$ printf 'one\ttwo\n'
one	two
```

Incorrect output from Ghostty (replaces a tab character with spaces):

```
$ printf 'one\ttwo\n'
one     two
```

If you're viewing this discussion in your browser, you can see the difference by highlighting the example output with your mouse. With spaces characters, you can highlight each space individually, whereas, with the tab character, you can only highlight the whole tab character.

**Reply by pe** [reply]
Same issue here. I have tools, that print tab separated values for copying to Excel. Works in iTerm and Terminal.app. Not in Ghostty.

**Reply by Oloremo** [reply]
Same here

**Reply by jacwah** [reply]
I have also used the same workflow for a long time with Kitty on macOS. I wrote many scripts that output TSV which I copy into Google Sheets. Since switching over to Ghostty this has become a hassle. I hope it can be fixed!

**Reply by brantfaircloth** [reply]
Same issue for me.  I generally expect output to the terminal that contains tabs to maintain those tabs. This is particularly important when copying terminal output to other programs (e.g. Excel or Google Sheets as mentioned above).

**Reply by cmonroe** [reply]
Same for me. I often work with Linux kernel patches and tabs need to be maintained. This is the only thing I have to go back to macOS Terminal for.

So long as I copy the data from a macOS Terminal window, I can paste it into Ghostty without losing tabs.

**Comment by johnnybubonic** [comment]
This has probably been the most annoying part of trying to use Ghostty. I switched to Ghostty for Zellij because spacing gets mucked in Zellij when copying because VTE doesn't implement OSC 52, only to find out that Ghostty doesn't properly render `\t` so... my copying of whitespace is still mucked. Awesome.

Here is what happens in Ghostty:
[ghostty.webm](https://github.com/user-attachments/assets/09bb5aac-ce35-436c-a06c-c3e8ee3d1471)

and here is the (objectively correct) behavior in Gnome Terminal:
[gnome-terminal.webm](https://github.com/user-attachments/assets/8005a5c1-8b2f-4512-ad36-bc42f919c12f)

**Reply by mitchellh** [reply]
Thanks, this behavior is already confirmed. I appreciate the extra videos demonstrating the issue, but your tone isn't appreciated. This a not-for-profit project run by a group of volunteers doing this for fun. If it bothers you so much, fix it yourself (or switch to another terminal). There's no room for being upset at the people dedicating their own free time to you. If I'm misreading your tone, I apologize, text is hard, but it came off that way.

**Reply by mitchellh** [reply]
> and here is the (objectively correct) behavior in Gnome Terminal:

I don't think it's objectively correct, fwiw. I agree with the behavior, but I don't know of any formal specifications where this specifies the behavior and xterm itself (the most objectively correct terminal emulator we really have) doesn't behave this way. It copies spaces. 

Like I said, I agree with the copy tabs behavior, but its not as black and white as it may seem to you.

**Reply by johnnybubonic** [reply]
> There's no room for being upset at the people dedicating their own free time to you.

I'm not upset, I'm just baffled that `0x09` / `U+0009` would be treated as one or more `0x20` / `U+0020`. They're completely different characters/bytes.

> I don't think it's objectively correct, fwiw.

Tabs are tabs (`0x09`), not a group of spaces (`0x20` * n).

> and xterm itself (the most objectively correct terminal emulator we really have) doesn't behave this way. It copies spaces.

XTerm directly acknowledges it *as a bug*.

It's the very first bulletpoint [in the "Known Bugs" page](https://invisible-island.net/xterm/xterm.faq.html#known_bugs), and also mentioned [in the FAQ](https://invisible-island.net/xterm/xterm.faq.html#xterm_tabs).

**Reply by mitchellh** [reply]
I've been testing various terminals and it's not yet clear yet what the desired behavior is, they all vary wildly. Here is a draft of my results so far:

## Testing Results

These are the results of `xxd` output of copying each test case in the script below.

| Terminal           | a                          | b                          | c                          | d                          | e                          |
| ------------------ | -------------------------- | -------------------------- | -------------------------- | -------------------------- | -------------------------- |
| Alacritty          | `6109 610a`                | `6220 2020 2020 2020 620a` | b                          | b                          | `6509 7820 2020 2065 0a`   |
| Gnome Console      | `6161 0a`                  | `6220 620a`                | `6320 2020 2020 2063 0a`   | `6420 2020 2020 2020 640a` | `6520 2078 650a`           |
| Kitty              | `6109 610a`                | `6220 2020 2020 2020 620a` | b                          | b                          | `6509 7820 2020 2065 0a`   |
| iTerm2             | `6109 610a`                | `6220 0962 0a`             | `6320 2020 2020 2009 630a` | `6420 2020 2020 2020 640a` | `6520 2078 0965`           |
| Terminal.app       | `6109 610a`                | `6220 2020 2020 2020 620a` | b                          | b                          | `6520 2078 2020 2020 650a` |
| XFCE               | `6109 610a`                | `6220 0962 0a`             | `6320 2020 2020 2009 630a` | `6420 2020 2020 2020 640a` | `6520 2078 0965`           |
| xterm              | `6120 2020 2020 2020 610a` | a                          | a                          | a                          | `6520 2078 2020 2020 650a` |
| Ghostty (5013d028) | `6120 2020 2020 2020 610a` | a                          | a                          | a                          | `6520 2078 2020 2020 650a` |

Ghostty currently matches xterm exactly, as designed. The interesting thing is the _wildly varying behavior_ of the other terminals on the non-obvious cases where intermediary characters within the tab spacing are replaced by printable characters. The only mostly consistent behavior is that all terminals emit a tab character when the tab is not overwritten at all (case "a"), ignoring Gnome Console and the xterm behavior.

## Verification Script

Here is a test script:

```sh
#!/bin/bash

# Clear the screen
printf "\033[2J"

# Move the cursor to the top left corner
printf "\033[H"

# Output: simple tab separated characters
printf "a\ta\n"

# Output: tab separated with a single space replaced by 0x20
printf "b\tb"
printf "\033[2;2H"
printf " "

# Output: replace all but one tabs with spaces
printf "\033[3;1H"
printf "c\tc"
printf "\033[3;2H"
printf "      "

# Output: replace all tabs with spaces
printf "\033[4;1H"
printf "d\td"
printf "\033[4;2H"
printf "       "

# Output: replace middle with a character
printf "\033[5;1H"
printf "e\te"
printf "\033[5;4H"
printf "x"

# Move cursor away
printf "\033[6;1H"
read
```

Copy any of the lines, and pipe to `xxd` to see their contents.

**Reply by mitchellh** [reply]
> XTerm directly acknowledges it as a bug.

In general, we also copy xterm's bugs unless there's an overwhelming reason to avoid it, because programs depend on them. In this case, I'm willing to diverge.

**Comment by daiyam** [comment]
> Like I said, I agree with the copy tabs behavior

@mitchellh Can you re-confirm it? Sorry about that, I got a bit burned about this issue with xterm.js...

For context:
- my xterm.js PR: https://github.com/xtermjs/xterm.js/pull/3434
- a discussion about the issue: https://github.com/xtermjs/xterm.js/issues/734

If you agree to the copy tabs behavior, I will be happy to implement it 😉

**Comment by mitchellh** [comment, marked as answer]
#8426
